presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPageSource.java

```diff
@@ -0,0 +1,242 @@
+            if (partitionKeys.containsKey(column.getId())) {
+                String partitionValue = partitionKeys.get(column.getId());
+                Type type = column.getType();
+                Object prefilledValue = deserializePartitionValue(type, partitionValue, column.getName(), timeZoneKey);
+                prefilledBlocks[outputIndex] = Utils.nativeValueToBlock(type, prefilledValue);
+                delegateIndexes[outputIndex] = -1;
+            }
+            else {
+                delegateIndexes[outputIndex] = delegateIndex;
+                delegateIndex++;
+            }
+            outputIndex++;
+        }
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return delegate.getCompletedBytes();
+    }
+
+    @Override
+    public long getCompletedPositions()
+    {
+        return delegate.getCompletedPositions();
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return delegate.getReadTimeNanos();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return delegate.isFinished();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        try {
+            Page dataPage = delegate.getNextPage();
+            if (dataPage == null) {
+                return null;
+            }
+            int batchSize = dataPage.getPositionCount();
+            Block[] blocks = new Block[prefilledBlocks.length];
+            for (int i = 0; i < prefilledBlocks.length; i++) {
+                if (prefilledBlocks[i] != null) {
+                    blocks[i] = new RunLengthEncodedBlock(prefilledBlocks[i], batchSize);
+                }
+                else {
+                    blocks[i] = dataPage.getBlock(delegateIndexes[i]);
+                }
+            }
+            return new Page(batchSize, blocks);
+        }
+        catch (RuntimeException e) {
+            closeWithSuppression(e);
+            throwIfInstanceOf(e, PrestoException.class);
+            throw new PrestoException(ICEBERG_BAD_DATA, e);
+        }
+    }
+
+    @Override
+    public void close()
+    {
+        try {
+            delegate.close();
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public String toString()
+    {
+        return delegate.toString();
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return delegate.getSystemMemoryUsage();
+    }
+
+    protected void closeWithSuppression(Throwable throwable)
+    {
+        requireNonNull(throwable, "throwable is null");
+        try {
+            close();
+        }
+        catch (RuntimeException e) {
+            // Self-suppression not permitted
+            if (throwable != e) {
+                throwable.addSuppressed(e);
+            }
+        }
+    }
+
+    private static Object deserializePartitionValue(Type type, String valueString, String name, TimeZoneKey timeZoneKey)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.